drivers/at86rf2xx/aes_spi.c

```diff
@@ -0,0 +1,356 @@
+/* Copyright (C) 2020 Otto-von-Guericke-Universität Magdeburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_at86rf2xx
+ * @{
+ *
+ * @file
+ * @brief       Implementation of at86rf2xx SPI security module (AES)
+ *
+ * @author      Fabian Hüßler <fabian.huessler@ovgu.de>
+ * @}
+ */
+
+#include "xtimer.h"
+#include "periph/spi.h"
+#include "at86rf2xx_aes.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#define AES_DEBUG(...)  DEBUG("[at86rf2xx_aes_spi]: "__VA_ARGS__)
+
+#define AT86RF2XX_CMD_SRAM_READ             (0b00000000)
+#define AT86RF2XX_CMD_SRAM_WRITE            (0b01000000)
+
+static inline
+void at86rf2xx_spi_get_bus(const at86rf2xx_t *dev)
+{
+    spi_acquire(dev->params.spi, dev->params.cs_pin, SPI_MODE_0, dev->params.spi_clk);
+}
+
+static inline
+void at86rf2xx_spi_release_bus(const at86rf2xx_t *dev)
+{
+    spi_release(dev->params.spi);
+}
+
+static inline
+uint8_t _aes_status(at86rf2xx_t *dev)
+{
+    spi_transfer_byte(dev->params.spi, dev->params.cs_pin, true,
+                      AT86RF2XX_CMD_SRAM_READ);
+    spi_transfer_byte(dev->params.spi, dev->params.cs_pin, true,
+                      AT86RF2XX_REG__AES_STATUS);
+    return spi_transfer_byte(dev->params.spi, dev->params.cs_pin, false, 0);
+}
+
+static inline
+uint8_t _aes_wait_for_result(at86rf2xx_t *dev)
+{
+    uint8_t status;
+    xtimer_usleep(AT86RF2XX_AES_DELAY_US);
+    if ((status = _aes_status(dev)) & AT86RF2XX_AES_STATUS_MASK__AES_ER) {
+        AES_DEBUG("ERROR\n");
+        return status;
+    }
+    while (!(status & AT86RF2XX_AES_STATUS_MASK__AES_DONE)) {
+        AES_DEBUG("status: %02x\n", status);
+        status = _aes_status(dev);
+    }
```

Comment:
Could we still get the `AT86RF2XX_AES_STATUS_MASK__AES_ER` bit set here?

Reply:
No the error bit is set before the transceiver has processed a data block. There are two cases:
1. The delay between initiating an AES operation and sending the next `cfg` to `AT86RF2XX_REG__AES_CTRL` was too short. Meaning the transceiver did not have enough time to process the current block.
2. Less then 16 bytes of data have been sent to the transceiver. This never happens in the code, so this case cannot arrive.

Reply:
Better add that as a comment so in the future this question doesn't come up again :)

Reply:
Actually the first case should not arrive either because I am polling if the operation is complete.
I could make an `assert()` that the error bit is never set and change the return types of all functions to be `void` because that was the only thing that I expected that could go wrong. Every time I witnessed the error bit was set, there was a mistake in the code.
Or there may be something that I am not aware of and the current implementation would be a little safer.
But I don´t think so. Even when I reduce `AT86RF2XX_AES_DELAY_US`, the error is not set because the polling is safe.
So I would like to turn it into an `assert()`